fedmsg/tests/crypto/test_x509.py

```diff
@@ -0,0 +1,216 @@
+        self.config.pop('ssldir')
+        with self.assertRaises(ValueError):
+            self.sign({'my': 'message'}, **self.config)
+
+    def test_sign(self):
+        """Assert signing a message inserts the certificate and a signature."""
+        signed = self.sign({'my': 'message'}, **self.config)
+        self.assertTrue('signature' in signed)
+        self.assertTrue('certificate' in signed)
+
+    def test_sign_and_verify(self):
+        """Assert signed messages are verified."""
+        signed = self.sign({'my': 'message'}, **self.config)
+        self.assertTrue(self.validate(signed, **self.config))
+
+    def test_unsigned(self):
+        """Assert unsigned messages are *not* verified."""
+        self.assertFalse(self.validate({'my': 'message'}, **self.config))
+
+    def test_invalid_ca(self):
+        """Assert when the CA didn't sign the certificate, validation fails."""
+        self.config['ca_cert_cache'] = os.path.join(SSLDIR, 'badca.crt')
+
+        signed = self.sign({'my': 'message'}, **self.config)
+        self.assertFalse(self.validate(signed, **self.config))
+
+    def test_invalid_signing_cert(self):
+        """Assert when the certificate didn't sign the message, validation fails."""
+        signed = self.sign({'my': 'message'}, **self.config)
+        self.config['certname'] = 'shell-app02.phx2.fedoraproject.org'
+        other_signed = self.sign({'my': 'message'}, **self.config)
+
+        signed['certificate'] = other_signed['certificate']
+
+        self.assertFalse(self.validate(signed, **self.config))
+
+    def test_signed_by_revoked_cert(self):
+        """Assert when the certificate has been revoked, validation fails."""
+        self.config['certname'] = 'dummy-revoked'
+        signed = self.sign({'my': 'message'}, **self.config)
+
+        self.assertFalse(self.validate(signed, **self.config))
+
+    def test_invalid_message_signature(self):
+        """Assert when the signed digest doesn't match the message digest, validation fails."""
+        signed = self.sign({'message': 'so secure'}, **self.config)
+        signed['message'] = 'so insecure!'
+        self.assertFalse(self.validate(signed, **self.config))
+
+    def test_invalid_policy(self):
+        """Assert when the policy is invalid, False is returned."""
+        self.config['routing_policy'] = {'mytopic': ['Not shell-app01.phx2.fedoraproject.org']}
+        self.config['routing_nitpicky'] = True
+        signed = self.sign({'topic': 'mytopic', 'message': 'so secure'}, **self.config)
+        self.assertFalse(self.validate(signed, **self.config))
+
+    def test_valid_policy(self):
+        """Assert when the policy is valid, True is returned."""
+        self.config['routing_policy'] = {'mytopic': ['shell-app01.phx2.fedoraproject.org']}
+        self.config['routing_nitpicky'] = True
+        signed = self.sign({'topic': 'mytopic', 'message': 'so secure'}, **self.config)
+        self.assertTrue(self.validate(signed, **self.config))
```

Comment:
How about a ```test_signed_by_expired_cert``` and ```test_signed_by_expired_ca``` (bad name, but what I mean is the signing cert is signed by an expired CA) as well?